AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMCs/StormTracker.dfy

```diff
@@ -122,31 +129,59 @@ module {:options "/functionSyntax:4" }  StormTracker {
       && wrapped.ValidState()
     }
     var wrapped : LocalCMC.LocalCMC // the actual cache
-    var inFlight: MutableMap<seq<uint8>, Types.PositiveLong> // the time at which this key became in flight
-    var gracePeriod : Types.PositiveLong // seconds before expiration that we start putting things in flight
-    var graceInterval : Types.PositiveLong // minimum seconds before putting the same key in flight again
-    var fanOut : Types.PositiveLong // maximum keys in flight at one time
-    var inFlightTTL : Types.PositiveLong // maximum time before a key is no longer in flight
+    const inFlight: MutableMap<seq<uint8>, Types.PositiveLong> // the time at which this key became in flight
+    const gracePeriod : Types.PositiveLong // seconds before expiration that we start putting things in flight
+    const graceInterval : Types.PositiveLong // minimum seconds before putting the same key in flight again
+    const fanOut : Types.PositiveLong // maximum keys in flight at one time
+    const inFlightTTL : Types.PositiveLong // maximum time before a key is no longer in flight
     var lastPrune : Types.PositiveLong // timestamp of last call to PruneInFlight
-    var sleepMilli : Types.PositiveLong // how long to sleep, if we sleep
+    const sleepMilli : Types.PositiveLong // how long to sleep, if we sleep
 
     constructor(
-      cache: Types.StormTrackingCache
+      nameonly ghost cache: Types.StormTrackingCache,
+      nameonly entryCapacity: nat,
+      nameonly entryPruningTailSize: nat,
+      nameonly gracePeriod: Types.PositiveLong,
+      nameonly graceInterval: Types.PositiveLong,
+      nameonly fanOut: Types.PositiveLong,
+      nameonly inFlightTTL: Types.PositiveLong,
+      nameonly sleepMilli: Types.PositiveLong
```

Comment:
If you can, don't do this this here.
Instead, create an Internal Datatype.
This will help us evolve this in the future.

You can then pass this around.
```
datatype InternalStormTrackerInput = 
  | InternalStormTrackerInput (
      nameonly ghost cache: Types.StormTrackingCache,
      nameonly entryCapacity: nat,
      nameonly entryPruningTailSize: nat,
      nameonly gracePeriod: Types.PositiveLong,
      nameonly graceInterval: Types.PositiveLong,
      nameonly fanOut: Types.PositiveLong,
      nameonly inFlightTTL: Types.PositiveLong,
      nameonly sleepMilli: Types.PositiveLong
 )
```

Reply:
This is what `nameonly`  on the arguments does. By requiring this on the method level there is an implicit datatype that can be evolved the way an internal datatype can.

What is the advantage you see in this additional `datatype`?

Reply:
I can do the mutation in the constructor. This saves updating the tests!

Reply:
For me, it has helped because if I need to add or remove an argument,
or change it's type,
I only have to do it one place,
as compared to every place it is called from.

Additionally, I have been playing with adding methods to datatypes.

But your constructor solution is ideal.